### ISSUE TITLE
Large skins cause client lag spikes

### ISSUE THREAD

**Issue by Desour** [ddnet/ddnet#9032]
I've recently noticed some client lag spikes (i.e. one frame too very long). And it seems to be related to players entering (and leaving?) the game.

I should probably add that I've also recently set my community skins download url to the not so curated <https://ddstats.tw/skins/>.

If I open the tee skin settings tab, there's also lag spikes. (I have a bunch of star-ed skins.)

I've done [some profiling](https://github.com/Desour/ddnet/tree/tracy_tmp1_skins) and some large skins take a considerable large amount of time to load. E.g. `LoadSkin` takes 91 ms `samogg`, and 200 ms for `twitter_emoji`.
The graphics thread seems to spend a large amount of time in uploading textures and buffers. (I'm using the opengl3.3 backend. Switching to the vulkan one doesn't fix the issue.)

**Comment by Jupeyy** [comment]
Then I can just suggest to not use that website as skin download URL.
If you know the author of that website ask them to remove such skins. 4k skins are for 32k monitors, nothing else.

So I expect your to have a RTX 6090 and a 32k Monitor :P

The problem here is, what should DDNet do?
Forbid these skins? That kinda sucks IMHO, breaks forward compatibility.
Use OpenGL multi threaded? The OpenGL spec does not specify what multi threading is.
Even for vulkan the queue operation is shared with the rendering commands.

For some reason ppl think that 4K skins are something desirable. I guess that is bcs of the common terminology: "This game is 4K ready". But a Tee skin only fills a very small portion of the screen, so 4k pixels inside this small rect:
![image](https://github.com/user-attachments/assets/7e7e781f-ea89-45c7-8d00-726df1bc6dc9)

**Comment by furo321** [comment]
> I should probably add that I've also recently set my community skins download url to the not so curated https://ddstats.tw/skins/.

Hey, I have resized all skins to a maximum resolution of `512x256`. Hope that helps!

**Comment by Desour** [comment]
>Use OpenGL multi threaded? The OpenGL spec does not specify what multi threading is.
Even for vulkan the queue operation is shared with the rendering commands.

I'm no graphics expert, but AFAIK vulkan does not run the commands that you queue to it sequentially.
What one could do is copy the data to the staging buffers in a separate thread. Or spread the work over multiple client steps.

>For some reason ppl think that 4K skins are something desirable. I guess that is bcs of the common terminology: "This game is 4K ready". But a Tee skin only fills a very small portion of the screen, so 4k pixels inside this small rect

I think I've seen someone praise their high resolution skin because they could zoom into and look at the high detailed fur.
(Not that I deem it necessary.)

>Hey, I have resized all skins to a maximum resolution of `512x256`. Hope that helps!

Thanks! It *seems* to help.
I'll close the issue then in a few days if issue confirms to be gone, and if just not using too high resolution skins is the recommended solution.

Btw., fyi, furo, some skins (i.e. fool1, fool2, and twinbox) cause warnings now:
`"downloadedskins/fool1.6114.tmp" is not compatible with pnglite and cannot be loaded by old DDNet versions: color type unsupported`

**Comment by Jupeyy** [comment]
> I'm no graphics expert, but AFAIK vulkan does not run the commands that you queue to it sequentially. What one could do is copy the data to the staging buffers in a separate thread. Or spread the work over multiple client steps.

That's correct, in a rust version of the vulkan backend I even do smth similar.
Sadly you still need to lock & block the queue while submitting your work.
Additionally the question is still how fast the GPU's vulkan driver copies from staging to VRAM internally, bcs that is probably the most expensive operation here.

- I just dunno how much more complicated I want to make the backends.. You could probably only copy the textures in small chunks to have better frames.
- You could only upload 1 texture per frame etc.

> I think I've seen someone praise their high resolution skin because they could zoom into and look at the high detailed fur.

Sure, to be fair I think for demo rendering it can also make sense. But if you zoom into the grass in GTA V it also looks shit xDD
At some level this is still _just_ a game and not a offline renderer.

**Comment by Robyt3** [comment]
The situation should have been improved a lot by #10068. Other than that, as @Jupeyy said, you shouldn't use skin databases with huge skins. Also, using the Vulkan backend, if it works for you, should be faster for uploading textures than OpenGL.
